firmware/spiders/linksys.py

```diff
@@ -1,133 +1,161 @@
+            return
+
+        page = kwargs['page']
+
+        for product_url in response.xpath(self.xpath['product_urls_on_page']).extract():
+            yield Request(url=response.urljoin(product_url), callback=self.move_to_support_page)
+
+        # move to next page in product catalogue
+        next_page = page + 1
+        yield Request(url=f'{response.url.rpartition("=")[0]}={next_page}', cb_kwargs=dict(page=next_page))
+
+    def move_to_support_page(self, response: Response) -> Optional[Request]:
+        support_page_matches = self.regex['get_support_page'].findall(response.body.decode())
+        if len(support_page_matches) < 1:
+            return None
+        return Request(url=response.urljoin(f'/de/support-product?rnId={support_page_matches[0]}'),
+                       callback=self.move_to_download_page)
+
+    @classmethod
+    def extract_date_and_version(cls, response: Response) -> Tuple[Optional[str], Optional[str]]:
+        matches = response.xpath(cls.xpath['date_and_version']).extract()
+        if len(matches) < 2:
+            return None, None
+
+        firmware_version = matches[0].replace('Ver.', '')
+        release_date = matches[1].split(' ')[-1].replace('/', '-')
+        return firmware_version, release_date
+
+    def move_to_download_page(self, response: Response) -> Optional[Request]:
+        download_page_matches = response.xpath(self.xpath['get_download_page']).extract()
+        if len(download_page_matches) < 1:
+            return None
+
+        product_name_matches = response.xpath(self.xpath['product_name']).extract()
+        if len(product_name_matches) < 1:
+            return None
+
+        device_name = product_name_matches[0][4:]
+        return Request(url=response.urljoin(download_page_matches[0]), callback=self.parse_download_page,
+                       cb_kwargs=dict(device_name=device_name), dont_filter=True)
+
+    def parse_download_page(self, response: Response, device_name: str) -> Generator[FirmwareItem, None, None]:
+        download_matches = response.xpath(self.xpath['download_link']).extract()
+        if len(download_matches) < 1:
+            return
+
+        file_url = download_matches[0]
+        firmware_version, release_date = self.extract_date_and_version(response)
+        device_class = self.map_device_class(device_name)
+
+        meta_data = {
+            'vendor': 'Linksys',
+            'file_urls': [file_url],
+            'device_name': device_name,
+            'device_class': device_class,
+            'firmware_version': firmware_version.strip(),
+            'release_date': release_date,
+        }
+
+        yield from self.item_pipeline(meta_data)
+
+    @classmethod
+    def map_device_class(cls, device_name: str) -> str:
+        for identifiers, device_class in cls.device_classes.items():
+            if device_name.startswith(identifiers.shortcuts):
+                return device_class
+        return ''
```

Comment:
Could it be a problem that `extract_date_and_version` returns `None` if nothing was found and this function returns an empty string?

Reply:
Yeah you're right! `extract_date_and_version` should return a tuple of empty strings. This would've caused a exception when trying to construct `meta_data`, as the firmware version gets stripped. Fixed that one!